File: app/src/main/java/com/moviesshumtimes/tv/ui/library/MovieDetailSections.kt
package com.moviesshumtimes.tv.ui.library

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.moviesshumtimes.tv.data.plex.PlexOnDeckItem
import com.moviesshumtimes.tv.data.plex.PlexPerson
import com.moviesshumtimes.tv.data.plex.PlexReview
import com.moviesshumtimes.tv.data.plex.PlexServer
import com.moviesshumtimes.tv.data.plex.PlexImageUrl
import com.moviesshumtimes.tv.ui.common.ShumArtwork
import com.moviesshumtimes.tv.ui.kit.ShumCard
import com.moviesshumtimes.tv.ui.kit.ShumCardContainer
import com.moviesshumtimes.tv.ui.kit.ShumTypography
import com.moviesshumtimes.tv.ui.kit.Text
import com.moviesshumtimes.tv.ui.theme.AppOnSurfaceVariant
import com.moviesshumtimes.tv.ui.theme.AppSurfaceVariant
import com.moviesshumtimes.tv.ui.theme.AppWhite
import com.moviesshumtimes.tv.ui.theme.NeonPurpleGlow

@Composable
fun CastCrewRow(
    server: PlexServer,
    cast: List<PlexPerson>,
    crew: List<PlexPerson>,
    onSelectPerson: (PlexPerson) -> Unit,
) {
    val people = cast + crew
    if (people.isEmpty()) return
    Column {
        Text(
            text = "Cast & Crew",
            style = ShumTypography.titleLarge,
            modifier = Modifier.padding(start = 32.dp, bottom = 16.dp),
        )
        LazyRow(
            contentPadding = PaddingValues(horizontal = 32.dp),
            horizontalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            itemsIndexed(people, key = { _, person -> person.id ?: person.tag }) { _, person ->
                val subtitle = person.role ?: crewTitle(person, crew)
                CastMemberAvatar(
                    server = server,
                    person = person,
                    subtitle = subtitle,
                    onClick = { onSelectPerson(person) },
                )
            }
        }
    }
}

private fun crewTitle(person: PlexPerson, crew: List<PlexPerson>): String? =
    if (person in crew) "Crew" else null

@Composable
private fun CastMemberAvatar(
    server: PlexServer,
    person: PlexPerson,
    subtitle: String?,
    onClick: () -> Unit,
) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier.width(112.dp),
    ) {
        ShumCard(onClick = onClick, shape = CircleShape, modifier = Modifier.size(84.dp)) {
            if (person.thumb != null) {
                ShumArtwork(
                    model = PlexImageUrl.of(server, person.thumb),
                    contentDescription = person.tag,
                    modifier = Modifier.fillMaxSize(),
                )
            } else {
                Box(
                    modifier = Modifier.fillMaxSize().background(AppSurfaceVariant),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(person.tag.take(1).uppercase(), style = ShumTypography.titleMedium, color = AppWhite)
                }
            }
        }
        Text(
            text = person.tag,
            maxLines = 1,
            overflow = TextOverflow.Ellipsis,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = 10.dp),
        )
        if (subtitle != null) {
            Text(
                text = subtitle,
                color = AppOnSurfaceVariant,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                textAlign = TextAlign.Center,
                modifier = Modifier.padding(top = 3.dp),
            )
        }
    }
}

@Composable
fun RatingsReviewsSection(
    rating: Double?,
    audienceRating: Double?,
    ratingImage: String?,
    audienceRatingImage: String?,
    reviews: List<PlexReview>,
) {
    if (rating == null && audienceRating == null && reviews.isEmpty()) return
    Column {
        Row(
            verticalAlignment = Alignment.Bottom,
            horizontalArrangement = Arrangement.spacedBy(14.dp),
            modifier = Modifier.padding(start = 32.dp, bottom = 16.dp),
        ) {
            Text(text = "Ratings & Reviews", style = ShumTypography.titleLarge)
            if (reviews.isNotEmpty()) {
                Text(text = "${reviews.size} critic reviews", color = AppOnSurfaceVariant)
            }
        }
        Row(
            horizontalArrangement = Arrangement.spacedBy(16.dp),
            modifier = Modifier.padding(horizontal = 32.dp),
        ) {
            if (rating != null) RatingTile(percent = (rating * 10).toInt(), label = "Critics", image = ratingImage)
            if (audienceRating != null) {
                RatingTile(percent = (audienceRating * 10).toInt(), label = "Audience", image = audienceRatingImage)
            }
        }
        if (reviews.isNotEmpty()) {
            LazyRow(
                contentPadding = PaddingValues(horizontal = 32.dp, vertical = 14.dp),
                horizontalArrangement = Arrangement.spacedBy(16.dp),
                modifier = Modifier.padding(top = 4.dp),
            ) {
                itemsIndexed(reviews.take(3), key = { _, review -> review.tag }) { _, review -> ReviewCard(review) }
            }
        }
    }
}

@Composable
private fun RatingTile(percent: Int, label: String, image: String?) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(14.dp),
        modifier = Modifier
            .background(AppWhite.copy(alpha = 0.04f), RoundedCornerShape(8.dp))
            .padding(horizontal = 22.dp, vertical = 16.dp),
    ) {
        Box(
            modifier = Modifier.size(40.dp).background(ratingBadgeColor(image), RoundedCornerShape(6.dp)),
        )
        Column {
            Text(text = "$percent%", style = ShumTypography.headlineMedium)
            Text(text = label, color = AppOnSurfaceVariant)
        }
    }
}

private fun ratingBadgeColor(image: String?): Color {
    val suffix = image?.substringAfterLast('.')?.lowercase()
    return when (suffix) {
        "fresh", "ripe", "certified" -> Color(0xFF2E7D4F)
        "rotten", "spilled" -> Color(0xFF7D2E2E)
        else -> AppSurfaceVariant
    }
}

@Composable
private fun ReviewCard(review: PlexReview) {
    Column(
        modifier = Modifier
            .width(280.dp)
            .background(AppWhite.copy(alpha = 0.04f), RoundedCornerShape(8.dp))
            .padding(18.dp),
    ) {
        val source = review.source
        if (source != null) {
            Text(text = source.uppercase(), color = NeonPurpleGlow)
        }
        Text(
            text = review.text,
            maxLines = 4,
            overflow = TextOverflow.Ellipsis,
            modifier = Modifier.padding(top = 10.dp),
        )
    }
}

@Composable
fun PosterRow(title: String, items: List<PlexOnDeckItem>, server: PlexServer, onClick: (PlexOnDeckItem) -> Unit) {
    if (items.isEmpty()) return
    Column {
        Text(
            text = title,
            style = ShumTypography.titleLarge,
            modifier = Modifier.padding(start = 32.dp, top = 4.dp, bottom = 16.dp),
        )
        LazyRow(
            contentPadding = PaddingValues(horizontal = 32.dp),
            horizontalArrangement = Arrangement.spacedBy(18.dp),
        ) {
            itemsIndexed(items, key = { _, item -> item.ratingKey }) { _, item ->
                RelatedPoster(server = server, item = item, onClick = { onClick(item) })
            }
        }
    }
}

@Composable
private fun RelatedPoster(server: PlexServer, item: PlexOnDeckItem, onClick: () -> Unit) {
    ShumCardContainer(
        modifier = Modifier.width(132.dp),
        imageCard = { interactionSource ->
            ShumCard(
                onClick = onClick,
                interactionSource = interactionSource,
                modifier = Modifier.fillMaxWidth().aspectRatio(2f / 3f),
            ) {
                ShumArtwork(
                    model = PlexImageUrl.of(server, item.thumb),
                    contentDescription = item.title,
                    modifier = Modifier.fillMaxSize(),
                )
            }
        },
        title = {
            Text(
                text = item.title,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
                modifier = Modifier.padding(top = 10.dp),
            )
        },
    )
}

File: flutter/lib/screens/library/movie_detail_sections.dart
import 'package:flutter/widgets.dart';

import '../../data/plex/plex_image_url.dart';
import '../../data/plex/plex_models.dart';
import '../../kit/card.dart';
import '../../kit/text.dart';
import '../../theme/tokens.dart';
import '../../theme/typography.dart';
import '../common/artwork.dart';

/// Ports ui/library/MovieDetailSections.kt's `CastCrewRow`.
class CastCrewRow extends StatelessWidget {
  final PlexServer server;
  final List<PlexPerson> cast;
  final List<PlexPerson> crew;
  final ValueChanged<PlexPerson> onSelectPerson;

  const CastCrewRow({super.key, required this.server, required this.cast, required this.crew, required this.onSelectPerson});

  @override
  Widget build(BuildContext context) {
    final people = [...cast, ...crew];
    if (people.isEmpty) return const SizedBox.shrink();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        const Padding(
          padding: EdgeInsets.only(left: 32, bottom: 16),
          child: AppText('Cast & Crew', style: AppTypography.titleLarge),
        ),
        SizedBox(
          height: 150,
          child: ListView.separated(
            scrollDirection: Axis.horizontal,
            padding: const EdgeInsets.symmetric(horizontal: 32),
            itemCount: people.length,
            separatorBuilder: (context, index) => const SizedBox(width: 18),
            itemBuilder: (context, index) {
              final person = people[index];
              final subtitle = person.role ?? (crew.contains(person) ? 'Crew' : null);
              return _CastMemberAvatar(
                key: ValueKey(person.id ?? person.tag),
                server: server,
                person: person,
                subtitle: subtitle,
                onClick: () => onSelectPerson(person),
              );
            },
          ),
        ),
      ],
    );
  }
}

class _CastMemberAvatar extends StatelessWidget {
  final PlexServer server;
  final PlexPerson person;
  final String? subtitle;
  final VoidCallback onClick;

  const _CastMemberAvatar({super.key, required this.server, required this.person, this.subtitle, required this.onClick});

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 112,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.center,
        mainAxisSize: MainAxisSize.min,
        children: [
          SizedBox(
            width: 84,
            height: 84,
            child: AppCard(
              onClick: onClick,
              shape: const CircleBorder(),
              child: person.thumb != null
                  ? SizedBox.expand(child: Artwork(imageUrl: PlexImageUrl.of(server, person.thumb)))
                  : ColoredBox(
                      color: AppColors.surfaceVariant,
                      child: Center(
                        child: AppText(
                          person.tag.isNotEmpty ? person.tag[0].toUpperCase() : '?',
                          style: AppTypography.titleMedium,
                          color: AppColors.white,
                        ),
                      ),
                    ),
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: AppText(person.tag, maxLines: 1, overflow: TextOverflow.ellipsis, textAlign: TextAlign.center),
          ),
          if (subtitle != null)
            Padding(
              padding: const EdgeInsets.only(top: 3),
              child: AppText(subtitle!, color: AppColors.onSurfaceVariant, maxLines: 1, overflow: TextOverflow.ellipsis, textAlign: TextAlign.center),
            ),
        ],
      ),
    );
  }
}

/// Ports ui/library/MovieDetailSections.kt's `RatingsReviewsSection`.
class RatingsReviewsSection extends StatelessWidget {
  final double? rating;
  final double? audienceRating;
  final String? ratingImage;
  final String? audienceRatingImage;
  final List<PlexReview> reviews;

  const RatingsReviewsSection({
    super.key,
    this.rating,
    this.audienceRating,
    this.ratingImage,
    this.audienceRatingImage,
    this.reviews = const [],
  });

  @override
  Widget build(BuildContext context) {
    if (rating == null && audienceRating == null && reviews.isEmpty) return const SizedBox.shrink();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 32, bottom: 16),
          child: Row(
            crossAxisAlignment: CrossAxisAlignment.end,
            mainAxisSize: MainAxisSize.min,
            children: [
              const AppText('Ratings & Reviews', style: AppTypography.titleLarge),
              if (reviews.isNotEmpty) ...[
                const SizedBox(width: 14),
                AppText('${reviews.length} critic reviews', color: AppColors.onSurfaceVariant),
              ],
            ],
          ),
        ),
        Padding(
          padding: const EdgeInsets.symmetric(horizontal: 32),
          child: Row(
            mainAxisSize: MainAxisSize.min,
            children: [
              if (rating != null) _RatingTile(percent: (rating! * 10).toInt(), label: 'Critics', image: ratingImage),
              if (rating != null && audienceRating != null) const SizedBox(width: 16),
              if (audienceRating != null) _RatingTile(percent: (audienceRating! * 10).toInt(), label: 'Audience', image: audienceRatingImage),
            ],
          ),
        ),
        if (reviews.isNotEmpty)
          Padding(
            padding: const EdgeInsets.only(top: 4),
            child: SizedBox(
              height: 140,
              child: ListView.separated(
                scrollDirection: Axis.horizontal,
                padding: const EdgeInsets.symmetric(horizontal: 32, vertical: 14),
                itemCount: reviews.length > 3 ? 3 : reviews.length,
                separatorBuilder: (context, index) => const SizedBox(width: 16),
                itemBuilder: (context, index) => _ReviewCard(key: ValueKey(reviews[index].tag), review: reviews[index]),
              ),
            ),
          ),
      ],
    );
  }
}

class _RatingTile extends StatelessWidget {
  final int percent;
  final String label;
  final String? image;

  const _RatingTile({required this.percent, required this.label, this.image});

  Color _badgeColor() {
    final suffix = image?.split('.').last.toLowerCase();
    if (suffix == 'fresh' || suffix == 'ripe' || suffix == 'certified') return const Color(0xFF2E7D4F);
    if (suffix == 'rotten' || suffix == 'spilled') return const Color(0xFF7D2E2E);
    return AppColors.surfaceVariant;
  }

  @override
  Widget build(BuildContext context) {
    return AppCard(
      onClick: () {},
      child: Container(
        decoration: BoxDecoration(color: AppColors.white.withValues(alpha: 0.04), borderRadius: BorderRadius.circular(8)),
        padding: const EdgeInsets.symmetric(horizontal: 22, vertical: 16),
        child: Row(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisSize: MainAxisSize.min,
          children: [
            Container(width: 40, height: 40, decoration: BoxDecoration(color: _badgeColor(), borderRadius: BorderRadius.circular(6))),
            const SizedBox(width: 14),
            Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              mainAxisSize: MainAxisSize.min,
              children: [
                AppText('$percent%', style: AppTypography.headlineMedium),
                AppText(label, color: AppColors.onSurfaceVariant),
              ],
            ),
          ],
        ),
      ),
    );
  }
}

class _ReviewCard extends StatelessWidget {
  final PlexReview review;

  const _ReviewCard({super.key, required this.review});

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 280,
      child: AppCard(
        onClick: () {},
        child: Container(
          decoration: BoxDecoration(color: AppColors.white.withValues(alpha: 0.04), borderRadius: BorderRadius.circular(8)),
          padding: const EdgeInsets.all(18),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            mainAxisSize: MainAxisSize.min,
            children: [
              if (review.source != null) AppText(review.source!.toUpperCase(), color: AppColors.accentGlow),
              Padding(
                padding: const EdgeInsets.only(top: 10),
                child: AppText(review.text, maxLines: 4, overflow: TextOverflow.ellipsis),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

/// Ports ui/library/MovieDetailSections.kt's `PosterRow` — reused for
/// related-hub rows and co-star rows.
class PosterRow extends StatelessWidget {
  final String title;
  final List<PlexOnDeckItem> items;
  final PlexServer server;
  final ValueChanged<PlexOnDeckItem> onClick;

  const PosterRow({super.key, required this.title, required this.items, required this.server, required this.onClick});

  @override
  Widget build(BuildContext context) {
    if (items.isEmpty) return const SizedBox.shrink();

    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      mainAxisSize: MainAxisSize.min,
      children: [
        Padding(
          padding: const EdgeInsets.only(left: 32, top: 4, bottom: 16),
          child: AppText(title, style: AppTypography.titleLarge),
        ),
        SizedBox(
          height: 232,
          child: ListView.separated(
            scrollDirection: Axis.horizontal,
            padding: const EdgeInsets.symmetric(horizontal: 32),
            itemCount: items.length,
            separatorBuilder: (context, index) => const SizedBox(width: 18),
            itemBuilder: (context, index) {
              final item = items[index];
              return _RelatedPoster(key: ValueKey(item.ratingKey), server: server, item: item, onClick: () => onClick(item));
            },
          ),
        ),
      ],
    );
  }
}

class _RelatedPoster extends StatelessWidget {
  final PlexServer server;
  final PlexOnDeckItem item;
  final VoidCallback onClick;

  const _RelatedPoster({super.key, required this.server, required this.item, required this.onClick});

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      width: 132,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        mainAxisSize: MainAxisSize.min,
        children: [
          AspectRatio(
            aspectRatio: 2 / 3,
            child: AppCard(
              onClick: onClick,
              child: SizedBox.expand(child: Artwork(imageUrl: PlexImageUrl.of(server, item.thumb))),
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(top: 10),
            child: AppText(item.title, maxLines: 1, overflow: TextOverflow.ellipsis),
          ),
        ],
      ),
    );
  }
}
